src/parsec.rs

```diff
@@ -17,38 +17,110 @@ use crate::{
 use log::LogLevel;
 #[cfg(not(feature = "mock_parsec"))]
 use parsec as inner;
-use std::collections::{btree_map::Entry, BTreeMap};
+use std::{
+    collections::{btree_map::Entry, BTreeMap},
+    fmt, mem,
+};
 
 #[cfg(feature = "mock_parsec")]
 pub use crate::mock::parsec::{
-    init_mock, ConsensusMode, NetworkEvent, Observation, Proof, PublicId, SecretId,
+    init_mock, ConsensusMode, Error, NetworkEvent, Observation, Proof, PublicId, SecretId,
 };
 #[cfg(not(feature = "mock_parsec"))]
-pub use parsec::{ConsensusMode, NetworkEvent, Observation, Proof, PublicId, SecretId};
+pub use parsec::{ConsensusMode, Error, NetworkEvent, Observation, Proof, PublicId, SecretId};
 
 pub type Block = inner::Block<chain::NetworkEvent, id::PublicId>;
 pub type Parsec = inner::Parsec<chain::NetworkEvent, FullId>;
 pub type Request = inner::Request<chain::NetworkEvent, id::PublicId>;
 pub type Response = inner::Response<chain::NetworkEvent, id::PublicId>;
 
+// TODO: Needs to be determined from better data.
+#[cfg(not(feature = "mock_parsec"))]
+const PARSEC_SIZE_LIMIT: usize = 1_000_000_000;
+// Mock parsec request/responses are much smaller, so we need a lower limit.
+#[cfg(feature = "mock_parsec")]
+const PARSEC_SIZE_LIMIT: usize = 100;
+
+// Keep track of size in case we need to prune.
+#[derive(Default, Debug, PartialEq, Eq)]
+struct ParsecSizeCounter(usize);
+
+impl ParsecSizeCounter {
+    fn increase_size(&mut self, size: usize) {
+        self.0 += size;
+    }
+
+    fn needs_pruning(&self) -> bool {
+        self.0 > PARSEC_SIZE_LIMIT
+    }
+}
+
+impl fmt::Display for ParsecSizeCounter {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "size: {}", self.0)
+    }
+}
+
+struct ParsecAndCounter {
+    pub parsec: Parsec,
+    pub size_counter: ParsecSizeCounter,
+}
+
+impl ParsecAndCounter {
+    fn handle_request(
+        &mut self,
+        src: &id::PublicId,
+        req: Request,
+        log_ident: &LogIdent,
+    ) -> Result<Response, Error> {
+        self.count_size(mem::size_of_val(&req), log_ident);
```

Comment:
Is `mem::size_of_val` doing the right thing? i.e is it taking into account that there is heap memory here?

Reply:
Well unlike `size_of` it's runtime check, so you'd think so. It's not super clear from the docs though if it's done recursively...

Reply:
Seems to be for array size, but if you can write a test or a playground example we can see.

Reply:
Yes you're right: https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=0dc9cb45007defdb046830cf541d0dad

Reply:
Updated: https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=6b13a34e5e9b498c4d8260cafd28f3bd